lib/data/database_repository.dart

```diff
@@ -0,0 +1,157 @@
+  T fromMap(Map<String, dynamic> json);
+  Map<String, dynamic> mapOf(T object);
+}
+
+/// Base class that adds a key [id] to allow its implementation to be stored/identified in any database
+abstract class KeyStorable extends Equatable {
+  const KeyStorable({required this.id});
+  final String id;
+}
+
+//
+// DatabaseRepository implementation using `sembast`
+//
+class DatabaseRepositoryImpl implements DatabaseRepository {
+  DatabaseRepositoryImpl(this._db);
+
+  // `sembast` database instance
+  final Database _db;
+
+  @override
+  Future<void> put<T extends KeyStorable>({
+    required T object,
+    required JsonSerializer<T> serializer,
+    required DatabaseStore store,
+    bool shouldMerge = true,
+  }) async {
+    final storeMap = stringMapStoreFactory.store(store.key);
+    final deserializedObject = serializer.mapOf(object);
+
+    await storeMap.record(object.id).put(_db, deserializedObject, merge: shouldMerge);
+  }
+
+  @override
+  Future<void> removeObject({required String key, required DatabaseStore store}) async {
+    final storeMap = stringMapStoreFactory.store(store.key);
+    await storeMap.record(key).delete(_db);
+  }
+
+  @override
+  Future<T?> getObject<T extends KeyStorable>({
+    required String key,
+    required JsonSerializer<T> serializer,
+    required DatabaseStore store,
+  }) async {
+    final storeMap = stringMapStoreFactory.store(store.key);
+    final rawObject = await storeMap.record(key).get(_db);
+
+    if (rawObject != null) {
+      return serializer.fromMap(rawObject);
+    }
+
+    return null;
+  }
+
+  @override
+  Future<List<T>> getAll<T extends KeyStorable>({
+    required JsonSerializer<T> serializer,
+    required DatabaseStore store,
+  }) async {
+    final storeMap = stringMapStoreFactory.store(store.key);
+
+    final allRecords = await storeMap.find(_db);
+
+    return allRecords.map((record) => serializer.fromMap(record.value)).toList();
+  }
+
+  @override
+  Future<Stream<List<T>>> listenAll<T extends KeyStorable>({
+    required JsonSerializer<T> serializer,
+    required DatabaseStore store,
+  }) async {
+    final storeMap = stringMapStoreFactory.store(store.key);
+
+    final transformer = _snapshotSerializerTransformer(serializer);
+    return storeMap.query().onSnapshots(_db).transform(transformer);
+  }
+
+  /// Transforms a list of `sembast` snapshot records into a list of objects parsed by [serializer]
+  StreamTransformer<List<RecordSnapshot<String, Map<String, Object?>>>, List<T>>
+      _snapshotSerializerTransformer<T extends Object>(JsonSerializer<T> serializer) {
+    return StreamTransformer.fromHandlers(
+      handleData: (snapshots, sink) {
+        final transformedRecords = snapshots.map((record) => serializer.fromMap(record.value)).toList();
+        sink.add(transformedRecords);
+      },
+    );
+  }
+}
+
+extension StoreKeys on DatabaseStore {
+  @visibleForTesting
+  String get key {
+    switch (this) {
```

Comment:
For this case is necessary a treatment for `default` of switch case?
Maybe add a custom throw error for this scenario, what do you think ?

Reply:
This is not necessary because this project is null-safe, thus a switch case in a non-optional enum will always require all cases to be filled, and it will always match one of the enumerator cases

Reply:
Make sense, tks =)